nixos/tests/incus/container.nix

```
@@ -0,0 +1,113 @@
+import ../make-test-python.nix ({ pkgs, lib, ... } :
+
+let
+  releases = import ../../release.nix {
+    configuration = {
+      # Building documentation makes the test unnecessarily take a longer time:
+      documentation.enable = lib.mkForce false;
+
+      # Our tests require `grep` & friends:
+      environment.systemPackages = [ pkgs.busybox ];
+    };
+  };
+
+  container-image-metadata = releases.lxdContainerMeta.${pkgs.stdenv.hostPlatform.system};
+  container-image-rootfs = releases.lxdContainerImage.${pkgs.stdenv.hostPlatform.system};
+in
+{
+  name = "incus-container";
+
+  meta.maintainers = with lib.maintainers; [ adamcstephens ];
+
+  nodes.machine = { ... }: {
+    virtualisation = {
+      diskSize = 4096;
+
+      # Since we're testing `limits.cpu`, we've gotta have a known number of
+      # cores to lean on
+      cores = 2;
+
+      # Ditto, for `limits.memory`
+      memorySize = 512;
+
+      incus.enable = true;
+    };
+  };
+
+  testScript = ''
+    def instance_is_up(_) -> bool:
+      status, _ = machine.execute("incus exec container --disable-stdin --force-interactive /run/current-system/sw/bin/true")
+      return status == 0
+
+    machine.wait_for_unit("incus.service")
+
+    # no preseed should mean no service
+    machine.fail("systemctl status incus-preseed.service")
+
+    machine.succeed("incus admin init --minimal")
+
+    machine.succeed(
+        "incus image import ${container-image-metadata}/*/*.tar.xz ${container-image-rootfs}/*/*.tar.xz --alias nixos"
+    )
+
+    with subtest("Container can be managed"):
+        machine.succeed("incus launch nixos container")
+        with machine.nested("Waiting for instance to start and be usable"):
+          retry(instance_is_up)
+        machine.succeed("echo true | incus exec container /run/current-system/sw/bin/bash -")
+        machine.succeed("incus delete -f container")
+
+    with subtest("Container is mounted with lxcfs inside"):
+        machine.succeed("incus launch nixos container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        ## ---------- ##
+        ## limits.cpu ##
+
+        machine.succeed("incus config set container limits.cpu 1")
+        machine.succeed("incus restart container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "1"
+            == machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
+        )
```

Not really a big deal, but maybe it would be better to split this up, to make it a bit easier to read? We could add error messages as well.

```suggestion
        cpuinfo = machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
        assert cpuinfo == "1", "Wrong number of CPUs reported from /proc/cpuinfo"
```

---

I've cleaned these all up. I didn't write most of this, but do agree much of this could be improved.